Use Bolt transport connections for execution

MikaelSmith opened the pull request:
Implement support for SSH and WinRM using Bolt's transport connections. This gives us consistent behavior with Bolt for those transports.

Only implements SSH and WinRM because Docker can use the Wash Docker plugin, local is trivial, and remote is not really usable. PCP I hope to implement later, but the lack of streaming responses will require more work.

MikaelSmith committed:
Implement target exec with Bolt transports

Adds support for all transports Bolt supports (except `remote`, because
it's not obvious how that should operate). Switched SSH to it as well
for consistency, because there would've been some cases where Wash's SSH
functioned differently (`run-as`, `sudo-executable`, and `proxyjump`
were ignored).

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Quote bash args so they're correctly expanded

Without this, arguments like 'a b c' would get split into 'a' 'b' 'c'.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Reimplement SSH and WinRM using Bolt transport connections

This is getting into the depths of Bolt, so it may be fragile to
maintain. But skip using `Bolt::Transport::*` because they're thin
wrappers when running a command, and override `execute` in
`Bolt::Transport::*::Connection` to stream stdout/stderr. Reuses the
logic present for sudo, connection setup, etc.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Ensure boltdir config is used

Wash looks for config related to the plugin. When running the plugin
directly as a script, config is ignored.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Use Dockerfile with locale to avoid Travis CI error

Getting
```
Unexpected error output on Exec: bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
```

This appears to have been addressed previously with Bolt. Using the Bolt
spec/Dockerfile to address this issue. Also plan to use the extra users
for future tests.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on bolt.rb:
```diff
@@ -117,36 +117,55 @@ class Target < Wash::Entry
+      connection.execute(command, stdin: opts[:stdin] ? $stdin : nil)
+    ensure
+      begin
+        connection&.disconnect
```
Should this be in a SIGTERM handler as well?

MikaelSmith replied:
Ensure handles that.

ekinanp commented on transport_ssh.rb:
```diff
@@ -0,0 +1,65 @@
+
+class BoltSSH < Bolt::Transport::SSH::Connection
+  # override execute so it streams output and returns the exit code
+  def execute(command, stdin: nil)
```
Some of this code seems copy-pasted from https://github.com/puppetlabs/bolt/blob/master/lib/bolt/transport/ssh/connection.rb#L224 ? If so, might be helpful to link that as a reference.

ekinanp replied:
You know, given all the code here, maybe breaking consistency isn't so bad. If I was reading this code, I'd be happy seeing something like
```
if supports_ssh
  // Implementing exec for SSH sucks, so we're just going to use Wash's
  // transport package to do it for us. Unfortunately no support exists for
  // WinRM, so we'll have to implement our own version in exec.
  transport :ssh, ...
end
```

MikaelSmith replied:
It's not that much code, and it is pretty much copied from Bolt with minor changes.

ekinanp replied:
Ok, fair enough.

ekinanp commented on bolt.rb:
```diff
@@ -117,36 +117,55 @@ class Target < Wash::Entry
+    begin
+      connection.connect
+      # Returns exit code
+      connection.execute(command, stdin: opts[:stdin] ? $stdin : nil)
```
Nit: Why not `stdin: opts[:stdin] || nil` ?`opts[:stdin]` should contain `$stdin`.

MikaelSmith replied:
Oh, forgot it gets switched from a boolean to $stdin. It's not well documented in wash-ruby.

ekinanp reviewed (commented):
Looks good in general, just some small comments re: code.

Have a few questions though:

* How hard would it be to add a hacky `exec` for the Docker transport? Initial idea is an `ExecDocker` method in the `transport` package and then using the `transport` feature. I get that breaks consistency, but I think expecting people to use the Docker plugin is the type of move that could introduce "pain for no good reason", unless doing exec for Docker is really hard and something we don't have time for given other priorities. By pain, I mean that it not only breaks the Boltwash interactive experience, but it may also require some extra, frustrating effort for a user. For example, it could lead to something like

> What was this target's raw container ID? How did I figure that out? Oh maybe `meta` has that. Ok there it is. Now to the Docker plugin now.. Navigating through the "fs".. Wait maybe my issue's in another target. Ok back to Boltwash, where was I again? Ohh no, I guess it's not in this target. Maybe it's in this other container's logs... Hmm.. Let me cd into it.. Ugh, right. I need to go to the Docker plugin to browse this thing after I use `meta` to figure out its raw container ID, wow so much for the "we reuse your hierarchy" crap in the plugin's README

All it takes is one bad experience to piss someone off, but I could also be paranoid about this.

* You still planning on adding WinRM support to the `transport` package tomorrow/next week, or leaving it alone for now? Asking out of curiosity because I'm fine with what you've got here. Surprised at how much simpler it is than SSH though.

MikaelSmith commented:
It's probably not too bad to support Docker. I'll add it.

MikaelSmith commented:
Added a rough plan for Windows target support to https://github.com/puppetlabs/wash/issues/699.

MikaelSmith committed:
Add Docker transport

This works differently than the Bolt plugin, and if you're trying to
operate on Bolt targets that use Docker the added switch to find it in
the Docker plugin would be irritating.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
Ok, this is ready again.

ekinanp commented on transport_docker.rb:
```diff
@@ -0,0 +1,39 @@
+    env_hash['DOCKER_HOST'] = @docker_host unless @docker_host.nil?
+
+    in_r, in_w = IO.pipe
+    in_w.sync = true
```
What part of this streams to stdout/stderr?

MikaelSmith replied:
Spawn's default behavior is to stream to inherit file descriptors (stdin/stdout/stderr) from the parent.

ekinanp replied:
Ah ok. Thanks

ekinanp reviewed (approved):
Looks good, just had one question.
